enterprise/cli/workspaceproxy.go

```diff
@@ -0,0 +1,506 @@
+				func(data any) (any, error) {
+					response, ok := data.(codersdk.CreateWorkspaceProxyResponse)
+					if !ok {
+						return nil, xerrors.Errorf("unexpected type %T", data)
+					}
+					return []codersdk.CreateWorkspaceProxyResponse{response}, nil
+				}),
+		)
+	)
+
+	client := new(codersdk.Client)
+	cmd := &clibase.Cmd{
+		Use:   "register",
+		Short: "Register a workspace proxy",
+		Middleware: clibase.Chain(
+			clibase.RequireNArgs(0),
+			r.InitClient(client),
+		),
+		Handler: func(inv *clibase.Invocation) error {
+			ctx := inv.Context()
+			resp, err := client.CreateWorkspaceProxy(ctx, codersdk.CreateWorkspaceProxyRequest{
+				Name:             proxyName,
+				DisplayName:      displayName,
+				Icon:             proxyIcon,
+				URL:              proxyURL,
+				WildcardHostname: proxyWildcardHostname,
+			})
+			if err != nil {
+				return xerrors.Errorf("create workspace proxy: %w", err)
+			}
+
+			var output string
+			if onlyToken {
+				output = resp.ProxyToken
+			} else {
+				output, err = formatter.Format(ctx, resp)
+				if err != nil {
+					return err
+				}
+			}
+
+			_, err = fmt.Fprintln(inv.Stdout, output)
+			return err
+		},
+	}
+
+	formatter.AttachOptions(&cmd.Options)
+	cmd.Options.Add(
+		clibase.Option{
+			Flag:        "name",
+			Description: "Name of the proxy. This is used to identify the proxy.",
+			Value:       clibase.StringOf(&proxyName),
+		},
+		clibase.Option{
+			Flag:        "display-name",
+			Description: "Display of the proxy. If omitted, the name is reused as the display name.",
+			Value:       clibase.StringOf(&displayName),
+		},
+		clibase.Option{
+			Flag:        "icon",
+			Description: "Display icon of the proxy.",
+			Value:       clibase.StringOf(&proxyIcon),
+		},
+		clibase.Option{
+			Flag:        "access-url",
+			Description: "Access URL of the proxy.",
+			Value:       clibase.StringOf(&proxyURL),
+		},
+		clibase.Option{
+			Flag:        "wildcard-access-url",
+			Description: "(Optional) Access url of the proxy for subdomain based apps.",
+			Value:       clibase.StringOf(&proxyWildcardHostname),
+		},
+		clibase.Option{
+			Flag:        "only-token",
+			Description: "Only print the token. This is useful for scripting.",
+			Value:       clibase.BoolOf(&onlyToken),
+		},
+	)
+	return cmd
+}
+
+type closers []func()
+
+func (c closers) Close() {
+	for _, closeF := range c {
+		closeF()
+	}
+}
+
+func (c *closers) Add(f func()) {
+	*c = append(*c, f)
+}
+
+func (r *RootCmd) proxyServer() *clibase.Cmd {
+	var (
+		cfg = new(codersdk.DeploymentValues)
+		// Filter options for only relevant ones.
+		opts = cfg.Options().Filter(codersdk.IsExternalProxies)
+
+		externalProxyOptionGroup = clibase.Group{
+			Name: "External Workspace Proxy",
+			YAML: "externalWorkspaceProxy",
+		}
+		proxySessionToken clibase.String
+		primaryAccessURL  clibase.URL
+		appSecuritYKey    clibase.String
+	)
+	opts.Add(
```

Comment:
Can't these be added to the regular config struct, and then the primary server can filter them out?

Reply:
We talked on discord. I think the best move here is to autogenerate the Deployment struct from the Options. Then we can create all the options in 1 place, and filter them.

I will do this in another branch, this is _ok for now_ here just to get a cmd working.

Follow up PRs will make the workspace proxy creation/register experience better.